src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,294 @@
+                    compilationUnit.Externs.AddRange(namespaceDeclaration.Externs), 
+                    compilationUnit.Usings.AddRange(namespaceDeclaration.Usings), 
+                    compilationUnit.AttributeLists, 
+                    members,
+                    eofToken).WithAdditionalAnnotations(Formatter.Annotation);
+            }
+
+            // Change namespace name
+            return root.ReplaceNode(namespaceDeclaration, 
+                namespaceDeclaration.WithName(
+                    CreateNameSyntax(targetNamespaceParts, aliasQualifier: null, targetNamespaceParts.Length - 1)
+                    .WithTriviaFrom(namespaceDeclaration.Name)
+                    .WithAdditionalAnnotations(WarningAnnotation, Formatter.Annotation)));
+        }
+
+        private static string GetAliasQualifierOpt(SyntaxNode name)
+        {
+            while (true)
+            {
+                switch (name.Kind())
+                {
+                    case SyntaxKind.QualifiedName:
+                        name = ((QualifiedNameSyntax)name).Left;
+                        continue;
+                    case SyntaxKind.AliasQualifiedName:
+                        return ((AliasQualifiedNameSyntax)name).Alias.Identifier.ValueText;
+                }
+
+                return null;
+            }
+        }
+
+        private NameSyntax CreateNameSyntax(ImmutableArray<string> namespaceParts, string aliasQualifier, int index)
+        {
+            var part = namespaceParts[index].EscapeIdentifier();
+            var namePiece = SyntaxFactory.IdentifierName(part);
+
+            if (index == 0)
+            {
+                return aliasQualifier == null ? (NameSyntax)namePiece : SyntaxFactory.AliasQualifiedName(aliasQualifier, namePiece);
+            }
+            else
+            {
+                return SyntaxFactory.QualifiedName(CreateNameSyntax(namespaceParts, aliasQualifier, index - 1), namePiece);
+            }
+        }
+
+        /// <summary>
+        /// return trivia attached to namespace declaration. 
+        /// Leading trivia of the node and trivia around opening brace, as well as
+        /// trivia around closing brace are concatenated together respectively.
+        /// </summary>
+        private static (List<SyntaxTrivia> openingTrivia, List<SyntaxTrivia> closingTrivia) 
+            GetOpeningAndClosingTriviaOfNamespaceDeclaration(NamespaceDeclarationSyntax namespaceDeclaration)
+        {
+            var openingTrivia = namespaceDeclaration.GetLeadingTrivia().ToList();
+            openingTrivia.AddRange(namespaceDeclaration.OpenBraceToken.LeadingTrivia);
+            openingTrivia.AddRange(namespaceDeclaration.OpenBraceToken.TrailingTrivia);
+
+            var closingTrivia = namespaceDeclaration.CloseBraceToken.LeadingTrivia.ToList();
+            closingTrivia.AddRange(namespaceDeclaration.CloseBraceToken.TrailingTrivia);
+
+            return (openingTrivia, closingTrivia);
```

Comment:
nit: in general we avoid Lists.  They're both mutable (bad), and they waste space (minor, but not great).  Instead, our preferred pattern is 'ArrayBuilder.GetInstance()' + .ToImmutableAndFree.  This ensures immutability, as well as not wasting any excess space